carta/session.py

```diff
@@ -358,39 +359,58 @@ def set_starting_directory(self, path):
+        """Open or append a new image.
 
         Parameters
         ----------
         path : {0}
             The path to the image file, either relative to the session's current directory or an absolute path relative to the CARTA backend's root directory.
         hdu : {1}
             The HDU to select inside the file.
-        complex : {2}
-            Whether the image is complex. Set to ``False`` by default.
-        expression : {3}
-            Arithmetic expression to use if opening a complex-valued image. The default is :obj:`carta.constants.ArithmeticExpression.AMPLITUDE`.
+        append : {2}
+            Whether the image should be appended to existing images. By default this is ``False`` and any existing open images are closed.
         """
-        return Image.new(self, path, hdu, False, complex, expression)
+        path = self.resolve_file_path(path)
+        directory, file_name = posixpath.split(path)
+        return Image.new(self, directory, file_name, hdu, append, False)
 
-    @validate(String(), String(r"\d*"), Boolean(), Constant(ArithmeticExpression))
-    def append_image(self, path, hdu="", complex=False, expression=ArithmeticExpression.AMPLITUDE):
-        """Append a new image, keeping any existing images.
+    @validate(String(), Constant(ComplexExpression), Boolean())
+    def open_complex_image(self, path, expression=ComplexExpression.AMPLITUDE, append=False):
+        """Open or append a new complex-valued image.
 
         Parameters
         ----------
         path : {0}
-            The path to the image file, either relative to the session's current directory or an absolute path relative to the CARTA backend's root directory.
-        hdu : {1}
-            The HDU to select inside the file.
-        complex : {2}
-            Whether the image is complex. Set to ``False`` by default.
-        expression : {3}
-            Arithmetic expression to use if appending a complex-valued image. The default is :obj:`carta.constants.ArithmeticExpression.AMPLITUDE`.
-        """
-        return Image.new(self, path, hdu, True, complex, expression)
+            The path to the complex-valued image file, either relative to the session's current directory or an absolute path relative to the CARTA backend's root directory.
+        expression : {1}
+            The complex expression for opening a complex-valued image. The default is :obj:`carta.constants.ComplexExpression.AMPLITUDE`.
+        append : {2}
+            Whether the image should be appended to existing images. By default this is ``False`` and any existing open images are closed.
+        """
+        path = self.resolve_file_path(path)
+        directory, file_name = posixpath.split(path)
+        expression = f'{expression}("{file_name}")'
+        return Image.new(self, directory, expression, "", append, True)
+
+    @validate(String(), String(), Boolean())
+    def open_LEL_image(self, expression, directory=None, append=False):
```

Comment:
Why directory is necessary here? What is its purpose at the frontend? Can we derive it internally just like load_image? Having it is quite counter-intuitive in my point of view.

Reply:
If a user wants to write a LEL expression for several files in the same base directory, it should be possible for them to specify the directory separately and use just relative file names in the expression (at least that's how I believe the frontend function uses the directory; this will have to be tested). But we have made the parameter optional -- so the user no longer *has to* fill this in (if all the paths in the expression are absolute and it doesn't matter what the directory is).

In the function for loading a hypercube, the directory parameter *never* matters (as far as we have tested), which is why we removed that parameter entirely.

Reply:
in the frontend we can do 
`"dice_four.fits" + "/User/kswang/carta_image_pool/set_dice/dice_one.fits"` in a directory containing dice_four.fits (ie the first image)
but we cannot do 
`"/User/kswang/carta_image_pool/set_dice/dice_one.fits" + "dice_four.fits"` in the same folder. So Based on this, I think we totally can derive the directory internally.

Reply:
If you're asking why we can't derive it from a single path like in `open_image`, it's because we'd have to parse the entire LEL expression in the wrapper (we allow any expression, but we just treat it as a string which we pass to the frontend to be parsed).

Reply:
how about in the docstring we explicitly say the images in the expression need to be all with absolution paths or all in the current directory? Then we don't need directory from users. Mixed usages like the examples I made are considered as edge cases which may or may not work. My intention here is to eliminate unclear user input.

Reply:
what happens if we always set directory as pwd (does the frontend do so?)?

Reply:
> in the frontend we can do `"dice_four.fits" + "/User/kswang/carta_image_pool/set_dice/dice_one.fits"` in a directory containing dice_four.fits (ie the first image) but we cannot do `"/User/kswang/carta_image_pool/set_dice/dice_one.fits" + "dice_four.fits"` in the same folder. So based on this, I think we totally can derive the directory internally.

Hmm. Do we know *why* that doesn't work? Is it because the browser's directory changes to the location of the first file, and then the directory is wrong for resolving the second file? This seems like undesirable behaviour. This means that if there's a mix of absolute and relative paths the LEL expression may not work no matter what directory is specified, depending on the order.

We can always set the directory to the "current" directory. But why should we remove the user's ability to use `open_LEL_image("A.fits + B.fits + C.fits", "path/to/those/files/")`, which should work (since all files are in the same directory)?

I'm also not sure whether we can use the session's `pwd` (as I thought we could). In the functions which have a separate path per file, we ensure that we can bypass the GUI browser's changing starting directory by converting relative paths to absolute paths (relative to our `pwd`, which doesn't change unless the user changes it). But if the frontend function changes the directory while processing the LEL expression, and calculates relative file paths using that directory, that's going to be a problem. I'm going to need to look more closely at how LEL expressions are handled.

It may be possible for us to parse the expression and rewrite the paths -- if we did that, we could ensure that the paths are all absolute and the expression always works (regardless of the starting directory and the order). But I would have to see how complex the syntax is.

Reply:
Or, we rename path as abs_path, just to ask users to use absolute paths. In the scripting interface this is fine I think.

Reply:
I think in the scripting interface "pwd" is not trivial to users. They need to explicitly use pwd to see the path. On the GUI, it is trivial in the file browser. Guessing this is the root of the problem/question.

Reply:
Since it's the only way for the circumstance that **_user_** only provide **_file name_** in `expression`. `load_image` also need user to provide `path` (the combination of `directory` and `file_name`) to derive.

Reply:
Sorry, since the cleaning staff went to our office for cleaning and I was out for my late lunch thus just missed the whole discussion.

I think we still need that `directory` argument since:
> in the frontend we can do `"dice_four.fits" + "/User/kswang/carta_image_pool/set_dice/dice_one.fits"` in a directory containing dice_four.fits (ie the first image) but we cannot do `"/User/kswang/carta_image_pool/set_dice/dice_one.fits" + "dice_four.fits"` in the same folder. So based on this, I think we totally can derive the directory internally.

When you said _"in a directory containing dice_four.fits"_, you have actually **specified the `directory` argument with your file browser**, and hence the action of specifying `directory` argument is the exact standard action taken by the front-end UI usage.

Therefore, I think the current logic for `open_LEL_image` is reasonable: let user to decide whether to provide directory in `expression` argument or with `directory` argument (but mention not to mix two ways together). If user wants to provide directory in `expression` argument, we clarify explicitly in the description that the input directory should be absolute directory.

Reply:
I still want to investigate why the order matters (in the GUI browser) and if there's a way to fix this (in the wrapper or in both the GUI and the wrapper). It would be nice to fix this -- intuitively I would expect the directory to be used as the base directory for all file paths in the expression, and for the order not to matter.